File: python/scripts/week-03/demo-01-embed-retrieve-ground.py
"""
Demo 1: Embed → Retrieve → Ground

The full RAG loop, made inspectable:

  1. EMBED    — every chunk becomes a vector (a point in space)
  2. RETRIEVE — the query is embedded, the closest chunks are returned
  3. GROUND   — those SAME chunks are injected into the prompt, and the
                LLM must answer from them

The point of this demo is not just to see an answer — it is to see the
evidence. Every chunk that grounds the answer is printed with its source,
and the prompt the model actually receives is shown.

Run: python scripts/week-03/demo-01-embed-retrieve-ground.py
"""
import os
import re
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from src.rag import (
    DATA_DIR,
    EMBEDDING_MODEL,
    SYSTEM_PROMPT,
    embed_text,
    grounded_answer_from_chunks,
    index_documents,
    retrieve_with_sources,
)

QUESTION = "What endpoints does the payment API expose?"
K = 4

BORDER = "=" * 72


def pause(prompt: str = "  ⏸  Press Enter to continue… ") -> None:
    """Pause so the learner can predict before the reveal.

    Non-interactive runs (piped/CI stdin) skip the pause instead of hanging.
    """
    try:
        input(prompt)
    except EOFError:
        print()


print(BORDER)
print("  Demo 1: Embed → Retrieve → Ground")
print(BORDER)
print()
print("  The RAG loop, step by step:")
print("    1. EMBED    — each chunk → a vector (a point in space)")
print(f"                   model: {EMBEDDING_MODEL} (384-dim vectors)")
print("    2. RETRIEVE — embed the query, return the closest chunks")
print("    3. GROUND   — inject those chunks into the prompt, answer")
print()
print(f"  Question: {QUESTION}")
print()

# ── Step 1: Embed + index ─────────────────────────────────────
print("  ── Step 1: Embed & index ────────────────────────────────")
print()
print("  Each document is split into chunks. Each chunk is embedded")
print("  into a vector and stored in Qdrant.")
print()
count = index_documents(chunk_size=512, chunk_overlap=64)
print(f"  ✅ Indexed {count} chunks → collection 'devbuddy-docs'")
print()

# ── Step 2: Retrieve ──────────────────────────────────────────
print("  ── Step 2: Retrieve (top-k by vector similarity) ────────")
print()
print("  The request — the query — is also embedded into a vector:")
print()
print(f'    "{QUESTION}"')
query_vector = embed_text(QUESTION)
print(f"    → {len(query_vector)}-dim vector: "
      f"{[round(v, 4) for v in query_vector[:5]]} ...")
print()
print(f"  Qdrant returns the {K} chunks whose vectors are closest to")
print("  that query vector, most relevant first. The source and similarity")
print("  score of each chunk are shown so you can trace it back:")
print()
chunks = retrieve_with_sources(QUESTION, k=K)
for i, chunk in enumerate(chunks, 1):
    print(f"  [{i}] {chunk.source}  (score={chunk.score:.3f})")
    print(f"      {chunk.content}")
    print()

print("  ⏸  PAUSE & PREDICT:")
print("     The payment API spec actually defines 4 endpoints. Look at the")
print(f"     {K} chunks above — which endpoint is MISSING from the retrieval?")
print("     Will the answer mention it? Predict, then continue.")
pause()

# ── Step 3: Ground ────────────────────────────────────────────
print("  ── Step 3: Ground (chunks → prompt → LLM) ───────────────")
print()
print("  The chunks above are injected VERBATIM into the system prompt")
print("  as CONTEXT. This is the exact prompt the model receives:")
print()
prompt_preview = SYSTEM_PROMPT.format(
    context="\n\n---\n\n".join(c.content for c in chunks)
)
for line in prompt_preview.splitlines():
    print(f"  │ {line}")
print()
print("  ── user message ──")
print(f"  │ {QUESTION}")
print()

answer = grounded_answer_from_chunks(QUESTION, [c.content for c in chunks], temperature=0.0)
print("  ── answer ──")
print(f"  │ {answer}")
print()

# ── Verify ────────────────────────────────────────────────────
print(BORDER)
print("  Verify the grounding (automated):")
context = "\n".join(c.content for c in chunks)
claims = re.findall(r"\b(?:GET|POST|PUT|DELETE|PATCH)\s+(/[^\s,`*]+)", answer)
if not claims:
    print("  ⚠️  No endpoint paths detected in the answer — verify by hand.")
else:
    ungrounded = [path for path in claims if path.rstrip(".") not in context]
    if ungrounded:
        print(f"  ❌ {len(ungrounded)}/{len(claims)} endpoint claim(s) NOT in retrieved context:")
        for path in ungrounded:
            print(f"      - {path}")
    else:
        print(f"  ✅ All {len(claims)} endpoint claim(s) appear in the retrieved context.")

# ── Recall check ──────────────────────────────────────────────
print()
print("  Recall check — grounding ≠ completeness:")
spec_path = os.path.join(DATA_DIR, "payment-api-spec.md")
with open(spec_path) as f:
    spec_text = f.read()
all_endpoints = re.findall(r"\b(?:GET|POST|PUT|DELETE|PATCH)\s+(/[^\s,`*]+)", spec_text)
missing = [e for e in all_endpoints if e.rstrip(".") not in context]
if missing:
    print(f"  ⚠️  {len(missing)} of {len(all_endpoints)} spec endpoint(s) were never retrieved:")
    for e in missing:
        print(f"      - {e}")
    print("     The model didn't hallucinate — it never SAW these. This is a")
    print("     RECALL gap (retrieval missed evidence), not a grounding failure")
    print("     (answer claims evidence that isn't there). Different bug, different fix.")
else:
    print(f"  ✅ All {len(all_endpoints)} spec endpoint(s) were retrieved.")
print(BORDER)

File: nodejs/scripts/week-03/demo-01-embed-retrieve-ground.js
/**
 * Demo 1: Embed → Retrieve → Ground
 *
 * The full RAG loop, made inspectable:
 *
 *   1. EMBED    — every chunk becomes a vector (a point in space)
 *   2. RETRIEVE — the query is embedded, the closest chunks are returned
 *   3. GROUND   — those SAME chunks are injected into the prompt, and the
 *                 LLM must answer from them
 *
 * The point of this demo is not just to see an answer — it is to see the
 * evidence. Every chunk that grounds the answer is printed with its source
 * and score, the prompt the model actually receives is shown, and two checks
 * separate grounding (precision) from recall (completeness).
 *
 * Run: node scripts/week-03/demo-01-embed-retrieve-ground.js
 */
import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import readline from "readline/promises";
import { stdin as input, stdout as output } from "process";
import {
  EMBEDDING_MODEL,
  SYSTEM_PROMPT,
  embedText,
  groundedAnswerFromChunks,
  indexDocuments,
  retrieveWithSources,
} from "../../src/rag.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const SPEC_PATH = resolve(__dirname, "../../../shared/data/payment-api-spec.md");

const QUESTION = "What endpoints does the payment API expose?";
const K = 4;
const BORDER = "=".repeat(72);

async function pause(msg = "  ⏸  Press Enter to continue… ") {
  if (!process.stdin.isTTY) return;
  const rl = readline.createInterface({ input, output });
  await rl.question(msg);
  rl.close();
}

console.log(BORDER);
console.log("  Demo 1: Embed → Retrieve → Ground");
console.log(BORDER);
console.log();
console.log("  The RAG loop, step by step:");
console.log("    1. EMBED    — each chunk → a vector (a point in space)");
console.log(`                   model: ${EMBEDDING_MODEL} (384-dim vectors)`);
console.log("    2. RETRIEVE — embed the query, return the closest chunks");
console.log("    3. GROUND   — inject those chunks into the prompt, answer");
console.log();
console.log(`  Question: ${QUESTION}`);
console.log();

// ── Step 1: Embed + index ─────────────────────────────────────
console.log("  ── Step 1: Embed & index ────────────────────────────────");
console.log();
console.log("  Each document is split into chunks. Each chunk is embedded");
console.log("  into a vector and stored in Qdrant.");
console.log();
const count = await indexDocuments(null, 512, 64);
console.log(`  ✅ Indexed ${count} chunks → collection 'devbuddy-docs'`);
console.log();

// ── Step 2: Retrieve ──────────────────────────────────────────
console.log("  ── Step 2: Retrieve (top-k by vector similarity) ────────");
console.log();
console.log("  The request — the query — is also embedded into a vector:");
console.log();
console.log(`    "${QUESTION}"`);
const queryVector = await embedText(QUESTION);
const preview = queryVector
  .slice(0, 5)
  .map((v) => Math.round(v * 10000) / 10000);
console.log(
  `    → ${queryVector.length}-dim vector: ${JSON.stringify(preview)} ...`
);
console.log();
console.log(`  Qdrant returns the ${K} chunks whose vectors are closest to`);
console.log("  that query vector, most relevant first. The source and similarity");
console.log("  score of each chunk are shown so you can trace it back:");
console.log();
const chunks = await retrieveWithSources(QUESTION, K);
chunks.forEach((chunk, i) => {
  console.log(`  [${i + 1}] ${chunk.source}  (score=${chunk.score.toFixed(3)})`);
  console.log(`      ${chunk.content}`);
  console.log();
});

console.log("  ⏸  PAUSE & PREDICT:");
console.log("     The payment API spec actually defines 4 endpoints. Look at the");
console.log(`     ${K} chunks above — which endpoint is MISSING from the retrieval?`);
console.log("     Will the answer mention it? Predict, then continue.");
await pause();
console.log();

// ── Step 3: Ground ────────────────────────────────────────────
console.log("  ── Step 3: Ground (chunks → prompt → LLM) ───────────────");
console.log();
console.log("  The chunks above are injected VERBATIM into the system prompt");
console.log("  as CONTEXT. This is the exact prompt the model receives:");
console.log();
const promptPreview = SYSTEM_PROMPT.replace(
  "{context}",
  chunks.map((c) => c.content).join("\n\n---\n\n")
);
for (const line of promptPreview.split("\n")) {
  console.log(`  │ ${line}`);
}
console.log();
console.log("  ── user message ──");
console.log(`  │ ${QUESTION}`);
console.log();

const answer = await groundedAnswerFromChunks(
  QUESTION,
  chunks.map((c) => c.content),
  0.0
);
console.log("  ── answer ──");
console.log(`  │ ${answer}`);
console.log();

// ── Verify ────────────────────────────────────────────────────
const context = chunks.map((c) => c.content).join("\n");
const claims = [
  ...answer.matchAll(/\b(?:GET|POST|PUT|DELETE|PATCH)\s+(\/[^\s,`*]+)/g),
].map((m) => m[1]);

console.log(BORDER);
console.log("  Verify the grounding (automated):");
if (claims.length === 0) {
  console.log("  ⚠️  No endpoint paths detected in the answer — verify by hand.");
} else {
  const ungrounded = claims.filter((p) => !context.includes(p.replace(/\.$/, "")));
  if (ungrounded.length) {
    console.log(
      `  ❌ ${ungrounded.length}/${claims.length} endpoint claim(s) NOT in retrieved context:`
    );
    ungrounded.forEach((p) => console.log(`      - ${p}`));
  } else {
    console.log(
      `  ✅ All ${claims.length} endpoint claim(s) appear in the retrieved context.`
    );
  }
}

// ── Recall check ──────────────────────────────────────────────
const specText = readFileSync(SPEC_PATH, "utf-8");
const allEndpoints = [
  ...specText.matchAll(/\b(?:GET|POST|PUT|DELETE|PATCH)\s+(\/[^\s,`*]+)/g),
].map((m) => m[1]);
const missing = allEndpoints.filter((e) => !context.includes(e.replace(/\.$/, "")));

console.log();
console.log("  Recall check — grounding ≠ completeness:");
if (missing.length) {
  console.log(
    `  ⚠️  ${missing.length} of ${allEndpoints.length} spec endpoint(s) were never retrieved:`
  );
  missing.forEach((e) => console.log(`      - ${e}`));
  console.log("     The model didn't hallucinate — it never SAW these. This is a");
  console.log("     RECALL gap (retrieval missed evidence), not a grounding failure");
  console.log("     (answer claims evidence that isn't there). Different bug, different fix.");
} else {
  console.log(`  ✅ All ${allEndpoints.length} spec endpoint(s) were retrieved.`);
}
console.log(BORDER);
